File: python/maze.py
from math import floor
from random import random

class Cell:

    wall_pairs = {
        'N': 'S', 
        'S': 'N', 
        'E': 'W', 
        'W': 'E', 
    }

    def __init__(self, x, y) -> None:
        self.x = x
        self.y = y
        self.walls = {
            'N': True,
            'S': True,
            'E': True,
            'W': True,
        }

    def has_all_walls(self):
        return all(self.walls.values())
    
    def knock_down_wall(self, other, wall):
        self.walls[wall] = False
        other.walls[Cell.wall_pairs[wall]] = False


def shuffle(array):
    for i in range(len(array)-1, -1, -1):
        j = floor(random() * (i + 1))

        temp = array[i]
        array[i] = array[j]
        array[j] = temp


class Maze:
    
    def __init__(self, nx, ny, ix=0, iy=0) -> None:
        self.nx = nx
        self.ny = ny
        self.ix = ix
        self.iy = iy
        self.maze_map = [[Cell(x, y) for y in range(ny)] for x in range(nx)]

    def cell_at(self, x, y):
        return self.maze_map[x][y]
    
    def print_maze(self):
        maze_rows = ["-" * (self.nx * 2)]
        
        for y in range(0, self.ny):
            maze_row = ["|"]

            for x in range(0, self.nx):
                if self.maze_map[x][y].walls["E"]:
                    maze_row.append(" |")
                else:
                    maze_row.append("  ")

            maze_rows.append("".join(maze_row))
            maze_row = ["|"]

            for x in range(0, self.nx):
                if self.maze_map[x][y].walls["S"]:
                    maze_row.append("—+")
                else:
                    maze_row.append(" +")
                
            maze_rows.append("".join(maze_row))
        
        return "\n".join(maze_rows)
    
    def find_valid_neighbours(self, cell):
        delta = [
            ["W", [-1, 0]],
            ["E", [1, 0]],
            ["S", [0, 1]],
            ["N", [0, -1]],
        ]
        neighbours = []

        for [direction, [dx, dy]] in delta:
            [x2, y2] = [cell.x + dx, cell.y + dy]

            if x2 >= 0 and x2 < self.nx and y2 >= 0 and y2 < self.ny:
                neighbour = self.cell_at(x2, y2)
                if neighbour.has_all_walls():
                    neighbours.append([direction, neighbour])
        
        return neighbours
    
    def make_maze(self):
        DIRECTIONS = {
            "N": [0, -1],
            "S": [0, 1],
            "E": [1, 0],
            "W": [-1, 0], 
        }
        OPPOSITE = {
            "N": "S",
            "S": "N",
            "E": "W",
            "W": "E", 
        }

        n = self.nx * self.ny
        cell_stack = []
        current_cell = self.cell_at(self.ix, self.iy)
        nv = 1

        while nv < n:
            neighbours = self.find_valid_neighbours(current_cell)
            if len(neighbours) == 0:
                current_cell = cell_stack.pop()
                continue

            [direction, next_cell] = neighbours[floor(random() * len(neighbours))]

            current_cell.knock_down_wall(next_cell, direction)
            cell_stack.append(current_cell)
            current_cell = next_cell
            nv += 1

        interior_cells = []
        for x in range(1, self.nx-1):
            for y in range(1, self.ny-1):
                interior_cells.append(self.cell_at(x, y))

        shuffle(interior_cells)

        num_walls_to_knock_down = round(len(interior_cells) * 4 * 0.1)

        for cell in interior_cells:
            walls_to_knock_down = []
            for direction in ["N", "S", "E", "W"]:
                if not cell.walls[direction]:
                    continue
                [dx, dy] = DIRECTIONS[direction]
                neighbor = self.cell_at(cell.x + dx, cell.y + dy)
                if not neighbor:
                    continue
                if not neighbor.has_all_walls():
                    walls_to_knock_down.append(direction)
            
            if len(walls_to_knock_down) > 0 and num_walls_to_knock_down > 0:
                direction = walls_to_knock_down[floor(random() * len(walls_to_knock_down))]
                cell.walls[direction] = False
                [dx, dy] = DIRECTIONS[direction]
                neighbor = self.cell_at(cell.x + dx, cell.y + dy)
                neighbor.walls[OPPOSITE[direction]] = False
                num_walls_to_knock_down -= 1

            if num_walls_to_knock_down == 0:
                break

File: OBSOLETE-js-prototype/maze.js
class Cell {
    static wall_pairs = {
        'N': 'S',
        'S': 'N',
        'E': 'W',
        'W': 'E'
    };

    constructor(x, y) {
        this.x = x;
        this.y = y;
        this.walls = {
            'N': true,
            'S': true,
            'E': true,
            'W': true
        };
    }

    has_all_walls() {
        return Object.values(this.walls).every(wall => wall === true);
    }

    knock_down_wall(other, wall) {
        this.walls[wall] = false;
        other.walls[Cell.wall_pairs[wall]] = false;
    }
}

function shuffle(array) {
    for (let i = array.length - 1; i > 0; i--) {
        const j = Math.floor(Math.random() * (i + 1));
        [array[i], array[j]] = [array[j], array[i]];
    }
}

export class Maze {
    constructor(nx, ny, ix = 0, iy = 0) {
        this.nx = nx;
        this.ny = ny;
        this.ix = ix;
        this.iy = iy;
        this.maze_map = Array.from(Array(nx), (col, x) =>
            Array.from(Array(ny), (cell, y) => new Cell(x, y))
        );
    }

    cell_at(x, y) {
        return this.maze_map[x][y];
    }

    toString() {
        let maze_rows = ["—".repeat(this.nx * 2)];
        for (let y = 0; y < this.ny; y++) {
            let maze_row = ["|"];
            for (let x = 0; x < this.nx; x++) {
                if (this.maze_map[x][y].walls["E"]) {
                    maze_row.push(" |");
                } else {
                    maze_row.push("  ");
                }
            }
            maze_rows.push(maze_row.join(""));
            maze_row = ["|"];
            for (let x = 0; x < this.nx; x++) {
                if (this.maze_map[x][y].walls["S"]) {
                    maze_row.push("—+");
                } else {
                    maze_row.push(" +");
                }
            }
            maze_rows.push(maze_row.join(""));
        }
        return maze_rows.join("\n");
    }

    find_valid_neighbours(cell) {
        const delta = [
            ["W", [-1, 0]],
            ["E", [1, 0]],
            ["S", [0, 1]],
            ["N", [0, -1]],
        ];
        let neighbours = [];
        for (let [direction, [dx, dy]] of delta) {
            let [x2, y2] = [cell.x + dx, cell.y + dy];
            if (x2 >= 0 && x2 < this.nx && y2 >= 0 && y2 < this.ny) {
                let neighbour = this.cell_at(x2, y2);
                if (neighbour.has_all_walls()) {
                    neighbours.push([direction, neighbour]);
                }
            }
        }
        return neighbours;
    }

    make_maze() {
        const DIRECTIONS = {
            "N": [0, -1],
            "S": [0, 1],
            "E": [1, 0],
            "W": [-1, 0]
        };
        
        const OPPOSITE = {
            "N": "S",
            "S": "N",
            "E": "W",
            "W": "E"
        };
        
        // Total number of cells.
        let n = this.nx * this.ny;
        let cell_stack = [];
        let current_cell = this.cell_at(this.ix, this.iy);
        // Total number of visited cells during maze construction.
        let nv = 1;

        while (nv < n) {
            let neighbours = this.find_valid_neighbours(current_cell);

            if (neighbours.length === 0) {
                // We've reached a dead end: backtrack.
                current_cell = cell_stack.pop();
                continue;
            }

            // Choose a random neighbouring cell and move to it.
            let [direction, next_cell] = neighbours[
                Math.floor(Math.random() * neighbours.length)
            ];
            current_cell.knock_down_wall(next_cell, direction);
            cell_stack.push(current_cell);
            current_cell = next_cell;
            nv += 1;
        }
        // Get a list of all the cells in the maze except for those on the edges.
        let interior_cells = [];
        for (let x = 1; x < this.nx - 1; x++) {
            for (let y = 1; y < this.ny - 1; y++) {
                interior_cells.push(this.cell_at(x, y));
            }
        }

        // Randomly shuffle the list of interior cells.
        shuffle(interior_cells);

        // Compute the number of walls to knock down (n% of the total number of interior walls).
        let num_walls_to_knock_down = Math.round(interior_cells.length * 4 * 0.1);

        // Loop over the interior cells and knock down walls until the desired number of walls has been knocked down.
        for (let cell of interior_cells) {
            let walls_to_knock_down = [];
            for (let direction of ["N", "S", "E", "W"]) {
                if (!cell.walls[direction]) {
                    continue;
                }
                let [dx, dy] = DIRECTIONS[direction];
                let neighbor = this.cell_at(cell.x + dx, cell.y + dy);
                if (!neighbor) {
                    continue;
                }
                if (!neighbor.has_all_walls()) {
                    walls_to_knock_down.push(direction);
                }
            }
            if (walls_to_knock_down.length > 0 && num_walls_to_knock_down > 0) {
                let direction = walls_to_knock_down[Math.floor(Math.random() * walls_to_knock_down.length)];
                cell.walls[direction] = false;
                let [dx, dy] = DIRECTIONS[direction];
                let neighbor = this.cell_at(cell.x + dx, cell.y + dy);
                neighbor.walls[OPPOSITE[direction]] = false;
                num_walls_to_knock_down--;
            }
            if (num_walls_to_knock_down == 0) {
                break;
            }
        }

    }
}